Fix errors from disallowed characters in BigQuery custom expression names

jeff303 opened the pull request:
Add `expression-name->alias` multimethod to sql.qp to handle translating the expression name to one that is supported by the driver

Add default impl that is essentially identity

Override for new BigQuery driver to run through the same `->valid-field-identifier` function already added for this purpose

Add test for a custom expression with weird characters

jeff303 committed:
Fix errors from disallowed characters in BigQuery custom expression names

Add `expression-name->alias` multimethod to sql.qp to handle translating the expression name to one that is supported by the driver

Add default impl that is essentially identity

Override for new BigQuery driver to run through the same `->valid-field-identifier` function already added for this purpose

Add test for a custom expression with weird characters

camsaul commented on src/metabase/driver/sql/query_processor.clj:
```diff
@@ -274,6 +274,18 @@
 
 (defmethod ->honeysql [:sql :value] [driver [_ value]] (->honeysql driver value))
 
+(defmulti ^String expression-name->alias
+  "Return the String alias that should be used for `expression-name`, to follow the `AS` clause. This is to allow for
+  translating names that particular databases may not allow as identifiers for custom expressions (even when quoted).
+  Defaults to identity (i.e. returns `expression-name` unchanged)."
+  {:added "0.41.0" :arglists '([driver expression-name])}
+  driver/dispatch-on-initialized-driver
+  :hierarchy #'driver/hierarchy)
```
I was actually considering adding something like this but more general, called `clause-alias`. Then we  can combine the existing `field->alias` method into it... 

https://github.com/metabase/metabase/pull/16656/files#diff-4fc60f4638cf98beb78946404682d4f857c629a070a01fc2d30112da04347b96R42-R46

A more general method might be better so we don't need to add `aggregation-reference->alias` or other stuff down the line

jeff303 replied:
So just to confirm, there is no chance that a driver might want different behavior for translating the name of an aggregation reference versus a field alias?

camsaul reviewed (approved):
This fix works for me 👍

The only thing I think we should think about is seeing whether we could generalize the identifier escaping method since we also have to do it for Field aliases (we already have `field->alias`). Maybe just a general `escape-identifier` method is a better solution? `expression-name->alias` and `field->alias` (once we've gone thru the deprecation cycle) can just be plain functions that call `escape-identifier`. This would also future proof us nicely since we want to support materialized views or other DDL stuff at some point in the long-term future.

LMK if that makes sense, if it's going to be too much of a hassle then it doesn't ruin my life to add the new separate method instead.

camsaul commented:
One more note: I'm 95% sure we have some sort of `escape-identifier` method of some sort in the driver test extension code... so we could also combine _that_ method into this one. So instead of a net gain of one new method this PR could actually be a net removal of one method

flamber commented:
May I just say - woohooo :tada: - probably 80 upvotes in #15008 will be very happy!
![image](https://user-images.githubusercontent.com/1447303/134769271-640e6bbb-0404-4487-984b-9e6a0fa7dfac.png)

jeff303 commented:
> One more note: I'm 95% sure we have some sort of `escape-identifier` method of some sort in the driver test extension code... so we could also combine _that_ method into this one. So instead of a net gain of one new method this PR could actually be a net removal of one method

Are you referring to [`qualified-name-components`](https://github.com/metabase/metabase/blob/master/test/metabase/test/data/sql.clj#L35)?

camsaul commented:
I don't think we need to worry about refactoring all of the test code to use the new `escape-identifier` method and do away with `format-name`, but it's at least worth marking `format-name` as `^:deprecated` and leave a note that we should use `escape-identifier` instead going forward

jeff303 commented:
@camsaul ,

I opted for `translate-identifier`, since I don't think "escaping" is quite accurate here (since you are actually renaming things, to some extent).  At any rate, I'm not sure we can really go to this level yet without figuring some things out.  For instance, H2 was doing `str/upper-case` to the "field or table" names, from its previous `format-name` (now `translate-identifier`).  The problem is applying that to expression aliases breaks a bunch of stuff (you can see the linked failures [here](https://app.circleci.com/pipelines/github/metabase/metabase/22235/workflows/b31e19f0-a83a-4123-8608-aa0044fed719/jobs/978100), for example).

I think we may actually need to keep the `expression-name->alias` multi (or something like it) to isolate that behavior somehow.  But I'm not sure what the best way is.  Thoughts?

camsaul commented:
That's a good point. I think maybe we should not try to subsume all of `format-name`'s responsibilities in that case, since it's doing something a little different. I think just the separate `escape-identifier`

jeff303 committed:
PR feedback

Change to a `escape-alias` multimethod, which by default simply returns the alias name unchanged

Override behavior for both BigQuery drivers to return a valid String

Removing `field->alias` multi and making it a regular (private) fn instead, which calls `escape-alias` with the field's `:name` (the BigQuery drivers were the only ones overriding it anyway)

camsaul commented on src/metabase/driver/sql/query_processor.clj:
```diff
@@ -181,18 +181,29 @@
-  "Return the string alias that should be used to for `field`, an instance of the Field model, i.e. in an `AS` clause.
-  The default implementation calls `:name`, which returns the *unqualified* name of the Field.
+(defmulti ^String escape-alias
+  "Return the String that should be emitted in the query for the given `alias-name`, which will follow the `AS` clause.
+  This is to allow for escaping names that particular databases may not allow as aliases for custom expressions
+  or fields (even when quoted).
 
-  Return `nil` to prevent `field` from being aliased."
-  {:arglists '([driver field])}
```
It might be a good idea to just leave this method unused for the time being and mark it as deprecated just so it's not a breaking change for 3rd-party drivers in 41. Normally we've been marking something as deprecated and removing it 3 versions later.

Not sure anybody else ever actually implemented this method but since it's not `^:private` we can't be sure

camsaul reviewed (approved):
LGTM. We should think about whether there's a way to leave `field->alias` in as a deprecated/unused method (maybe rename the function version to something else) to give 3rd party driver authors a chance to remove their impls if they have one

jeff303 committed:
PR feedback

Switch `field->alias` back to a (deprecated) multimethod

camsaul commented on src/metabase/driver/sql/query_processor.clj:
```diff
@@ -181,18 +181,36 @@
-  Return `nil` to prevent `field` from being aliased."
+  DEPRECATED as of x.41.  This multimethod will be removed in a future release.  Instead, drivers will simply
+  override the `escape-alias` multimethod if they want to influence the alias to be used for a given field name."
   {:arglists '([driver field])}
```
```suggestion
  {:arglists '([driver field]), :deprecated "0.41.0"}
```

jeff303 replied:
Fixed

camsaul commented on src/metabase/driver/sql/query_processor.clj:
```diff
@@ -181,18 +181,36 @@
+  [_ alias-name]
+  alias-name)
+
+(defmulti ^:deprecated ^String field->alias
```
```suggestion
(defmulti ^String field->alias
```

jeff303 replied:
Fixed

jeff303 committed:
PR feedback

Fix deprecation
